### ISSUE TITLE
Throw Error When popup closes without entering credentials.

### ISSUE THREAD

**Issue by royipressburger** [ronzeidman/ng2-ui-auth#75]
Currently (On devices) when the popup window is closes without entering credentials, (By click the close button or the back button on android) no event is being return.

Im poping up a loaded when users ask to login with facebook, If the users return from that window without entering the credentials I get no indication on it. Thus the loader stays always up.

We need to think how to grab that event and reject/throw error to the client.
(Im closing the pull request i have opened since it doesn't solve that problem)

**Comment by ronzeidman** [comment]
Can you try 6.1.0, I've thrown a `new Error('Authentication Canceled');` when the window closes.
all you need to do when authenticating is 
```ts
.subscribe(
    next => <do something>,
    error => <do something and stop spinner>,
    (/* complete */) => <stop spinner>
)
```

**Comment by royipressburger** [comment]
Cool.. ill test and let you know!

**Comment by royipressburger** [comment]
@ronzeidman I have updated to version 6.1.1, and it doesn't seems to work.
On web it is working ok but on a mobile device(with cordova inapp-browser), It looks like nothing is invoked when closing the window.. Ill try to debug more with android emulator and try to get back to you with a better response.

Thanks for your help!

**Comment by royipressburger** [comment]
@ronzeidman So I think I have found the issue.. and still looking for a solution..

[here](https://github.com/ronzeidman/ng2-ui-auth/blob/master/src/popup.service.ts#L78) you just taking the `load` event.. why?
I think here we miss the close event. could be?

**Comment by ronzeidman** [comment]
Seems so. Can you try adding `.merge(Observable.fromEvent<Event>(this.popupWindow, 'close').switchMap(() => Observable.throw(new Error ('Authentication Canceled'))))` before the switchMap?  I think that should do the trick.

**Comment by royipressburger** [comment]
@ronzeidman How do you build the project?
I have cloned and change my path of the module in the package.json. I ran `npm run compile` but im getting some weird error in my project that I didnt get from installing the package normally.

`Property 'map' does not exist on type 'Observable<Response>'. `

**Comment by royipressburger** [comment]
@ronzeidman I manage to run it only after pushing to git and then set my package json to git.
Ill be happy to know how to working with it locally :)

Any case... it looks like the right event is `exit`, but the issue now that it is grabbed very time the window closes.. hence there is no redirect... any ideas?

**Comment by ronzeidman** [comment]
Maybe add a timer? `.merge(Observable.fromEvent<Event>(this.popupWindow, 'exit').switchMap(() => Observable.timer(100).switchMap(() => Observable.throw(new Error ('Authentication Canceled')))))`
It's a hack but it should work.

**Comment by ronzeidman** [comment]
.delayTime maybe more appropriate. I'm not near  my computer so I can't check...

**Comment by ronzeidman** [comment]
You can use npm package locally by `npm link` or `npm i <folder name>`

**Comment by royipressburger** [comment]
I tried the link but still have the same issue. Also the delay doesn't work since it has to be longer then the normal auth request, putting a long timeout can work but thats :/.

Im still trying to think of something.

**Comment by ronzeidman** [comment]
The delay should be on the popup exit event. Then it 100 ms should be enough. The error should stop any other value from being emitted.

**Comment by ronzeidman** [comment]
Try adding the `merge` here:
```ts
 eventListener(redirectUri: string) {
        return Observable
            .fromEvent<Event>(this.popupWindow, 'loadstart')
            .switchMap((event: Event & { url: string }) => {
                if (!this.popupWindow || this.popupWindow.closed) {
                    return Observable.throw(new Error('Authentication Canceled'));
                }
                if (event.url.indexOf(redirectUri) !== 0) {
                    return Observable.empty();
                }

                let parser = document.createElement('a');
                parser.href = event.url;

                if (parser.search || parser.hash) {
                    const queryParams = parser.search.substring(1).replace(/\/$/, '');
                    const hashParams = parser.hash.substring(1).replace(/\/$/, '');
                    const hash = PopupService.parseQueryString(hashParams);
                    const qs = PopupService.parseQueryString(queryParams);
                    const allParams = assign({}, qs, hash);

                    this.popupWindow.close();

                    if (allParams.error) {
                        throw allParams.error;
                    } else {
                        return Observable.of(allParams);
                    }
                }
                return Observable.empty();
            })
            .merge(Observable.fromEvent<Event>(this.popupWindow, 'exit').delay(100).map(() => throw new Error('Authentication Canceled')))
            .take(1);
    }
```

**Comment by royipressburger** [comment]
hi. sorry for late response.. but im getting 
`Property 'merge' does not exist on type 'Observable<any>`

I can fix it by changing the import of observable from `rxjs/Observable` to `rxjs/Rx` but them im getting error on other places.

**Comment by royipressburger** [comment]
Its working now, There is a pull request #82

**Comment by ronzeidman** [comment]
Included in 6.1.2
